README.md

# NAME

Auth::OATH::OTP - OATH One-Time Passcode Generator

# SYNOPSIS

    use Auth::OATH::OTP;
    my $oath = Auth::OATH::OTP->new;
    my $totp = $oath->totp('MySecretPassword');
    my $hotp = $oath->hotp('MyOtherSecretPassword', 1);

# DESCRIPTION

Auth::OATH::OTP is implementation of the HOTP (RFC 4226) and TOTP (RFC 6238)
algorithms as defined by OATH. (http://www.openauthentication.org)

# METHODS

## __Auth::OATH::OTP->new(%options)__

    my $oath = Auth::OATH::OTP->new(
        algorithm => 'sha1',
        digits    =>  6,
        timestep  => 30,
    );

- __algorithm__

    Allowed values for _algorithm_ are 1, 224, 256, 384, 512, 512224, or 512256.
    For more information about this parameter, see also [Digest::SHA](http://search.cpan.org/perldoc?Digest::SHA).

    If the argument is missing, SHA-1 will be used by default.

- __digits__

    `totp( )` and `hotp( )` both default to returning 6 digits.

- __timestep__

    _timestep_ only applies to the `totp( )` function.

    By default, the _timestep_ is 30 seconds, so there is a new password every 30 seconds.

## __$oath->totp($secret\_key\[, $time\]) : String__

    my $secret = pack 'A*', 'My Secret Key';
    my $totp   = $oath->totp($secret);
    # or
    my $totp   = $oath->totp($secret, time);

Returns a text string with the Time-based One Time Passcode.

Manual _time_ is an optional parameter.
If it is missing, the current time will be used by default.
This is useful for testing purposes.

## __$oath->hotp($secret\_key, $counter) : String__

    my $secret = pack 'A*', 'My Secret Key';
    my $hotp   = $oath->hotp($secret, 1);

Returns a text string with the HMAC-based One Time Passcode.

Both parameters are required.

# LICENSE

Copyright (C) hatyuki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

hatyuki <hatyuki29@gmail.com>

# SEE ALSO

HOTP: [http://www.ietf.org/rfc/rfc4226.txt](http://www.ietf.org/rfc/rfc4226.txt)

TOTP: [http://www.ietf.org/rfc/rfc6238.txt](http://www.ietf.org/rfc/rfc6238.txt)

[Authen::OATH](http://search.cpan.org/perldoc?Authen::OATH)


Changes
Revision history for Perl extension Auth-OATH-OTP

{{$NEXT}}

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Auth-OATH-OTP',
    module_name     => 'Auth::OATH::OTP',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


benchmark/benchmark.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Auth::OATH::OTP;
use Authen::OATH;
use Benchmark qw/ cmpthese timethese /;

my $secret = pack 'A*', '12345678901234567890';
my $this = Auth::OATH::OTP->new;
my $that = Authen::OATH->new;

die if $this->totp($secret) ne $that->totp($secret);

cmpthese timethese 100000, +{
    'Auth::OATH::OTP' => sub {
        my $oath = Auth::OATH::OTP->new;
        $oath->totp($secret);
    },
    'Authen::OATH' => sub {
        my $oath = Authen::OATH->new;
        $oath->totp($secret);
    },
    'Auth::OATH::OTP (reuse)' => sub {
        $this->totp($secret);
    },
    'Authen::OATH (reuse)' => sub {
        $that->totp($secret);
    },
};

__END__

Auth::OATH::OTP:  4 wallclock secs ( 3.86 usr +  0.01 sys =  3.87 CPU) @ 25839.79/s (n=100000)
Auth::OATH::OTP (reuse):  4 wallclock secs ( 3.58 usr +  0.02 sys =  3.60 CPU) @ 27777.78/s (n=100000)
Authen::OATH: 33 wallclock secs (33.38 usr +  0.05 sys = 33.43 CPU) @ 2991.33/s (n=100000)
Authen::OATH (reuse): 12 wallclock secs (11.84 usr +  0.03 sys = 11.87 CPU) @ 8424.60/s (n=100000)
                           Rate Authen::OATH Authen::OATH (reuse) Auth::OATH::OTP Auth::OATH::OTP (reuse)
Authen::OATH             2991/s           --                 -64%            -88%                    -89%
Authen::OATH (reuse)     8425/s         182%                   --            -67%                    -70%
Auth::OATH::OTP         25840/s         764%                 207%              --                     -7%
Auth::OATH::OTP (reuse) 27778/s         829%                 230%              8%                      --


lib/Auth/OATH/OTP/Verifier.pm
package Auth::OATH::OTP::Verifier;
use strict;
use warnings;
use parent qw/ Auth::OATH::OTP /;
use Carp ( );
use Convert::Base32 ( );
use List::Util ( );
use URI::Escape ( );
use Class::Accessor::Lite (
    new => 0,
    ro  => [qw/ algorithm binkey digits label secret timestep types window /],
);

sub new
{
    my $class = shift;
    my %args  = scalar @_ == 1 && ref $_[0] eq 'HASH' ? %{ $_[0] } : @_;

    unless (defined $args{secret}) {
        $args{secret} = generate_secret_key(32);
    }

    return $class->SUPER::new(
        binkey => Convert::Base32::decode_base32($args{secret}),
        label  => '',
        types  => 'totp',
        window => 3,
        %args,
    );
}

sub generate_secret_key
{
    my $length = defined $_[0] && $_[0] > 0 ? $_[0] : 32;
    my @chars  = List::Util::shuffle('A'..'Z', 2..7);
    return join '', map { $chars[int rand 32] } 1..$length;
}

sub verify
{
    my ($self, $code, $seed) = @_;
    return 0 unless defined $code;

    my $method = lc $self->types;
    Carp::croak "Unknown algorithm '$method'" unless $self->can($method);

    unless (defined $seed) {
        if ($method eq 'totp') {
            $seed = time;
        } else {
            Carp::croak 'Required parameter missing: $counter';
        }
    }

    # HOTP
    if ($method eq 'hotp') {
        return $self->hotp($seed) == $code ? 1 : 0;
    }

    # TOTP
    my $window = $self->window;
    for (my $i = int -(($window - 1) / 2) ; $i < $window / 2 ; ++$i) {
        return 1 if $self->totp($seed + $i * $self->timestep) == $code;
    }

    return 0;
}

sub totp
{
    my ($self, $time) = @_;
    return $self->SUPER::totp($self->binkey, $time);
}

sub hotp
{
    my ($self, $counter) = @_;
    return $self->SUPER::hotp($self->binkey, $counter);
}

sub key_uri
{
    my $self   = shift;
    my $types  = lc $self->types;
    my $label  = URI::Escape::uri_escape_utf8($self->label);
    my $secret = 'secret=' . $self->secret;
    return sprintf 'otpauth://%s/%s?%s',$types, $label, $secret;
}

1;

__END__

=encoding utf-8

=head1 NAME

Auth::OATH::OTP::Verifier - One-Time Passcode Verifier

=head1 SYNOPSIS

    use Auth::OATH::OTP::Verifier;
    my $auth = Auth::OATH::OTP::Verifier->new;

    # Authentification
    if ($auth->verify('012345')) {
        # Success
    } else {
        # Fail
    }

    # Generate Time-based One Time Passcode
    my $totp_code = $auth->totp;

    # Generate HMAC-based One Time Passcode
    my $counter   = 15;
    my $hotp_code = $auth->hotp($counter);

=head1 DESCRIPTION

Auth::OATH::OTP::Verifier is ...

=head1 METHODS

=head2 B<< Auth::OATH::OTP::Verifier->new(%options) >>

    my $oath = Auth::OATH::OTP->new(
        algorithm => 'sha1',
        digits    =>  6,
        timestep  => 30,
        label     => 'My Authentification',
        types     => 'TOTP', # Lower-case letters are also allowed.
        window    => 3,
        secret    => 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567',
    );

=over

=item B<algorithm>

=item B<digits>

=item B<timestep>

If you want to know about these parameters, see L<Auth::OATH::OTP>.

=item B<label>

The label is used to identify which account a key is associated with.

=item B<types>

Valid I<types> are B<HOTP> and B<TOTP>,to distinguish whether the key will be
used for counter-based HOTP or for TOTP.

=item B<window>

If the argument is missing, 3 will be used by default.

=item B<secret>

The I<secret> parameter is an arbitrary secret key value encoded in Base32 according to RFC 3548.
If the argument is missing, random Base32 string of length 32 will be used by default.

=back

=head2 B<< $oath->verify($code[, $seed]) : Boolean >>

    my $totp_verifier   = Auth::OATH::OTP::Verifier->new(types => 'totp');
    my $totp_is_success = $totp_verifier->verify($code[, $time]);
    # or
    my $hotp_verifier   = Auth::OATH::OTP::Verifier->new(types => 'hotp');
    my $hotp_is_success = $hotp_verifier->verify($code, $counter]);

=head2 B<< $oath->totp([$time]) : String >>

    my $totp = $oath->totp;
    # or
    my $totp = $oath->totp(time);

Returns a text string with the Time-based One Time Passcode.

Manual I<time> is an optional parameter.
If it is missing, the current time will be used by default.
This is useful for testing purposes.

=head2 B<< $oath->hotp($counter) : String >>

    my $hotp = $oath->hotp(12345);

Returns a text string with the HMAC-based One Time Passcode.

=head1 LICENSE

Copyright (C) hatyuki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

hatyuki E<lt>hatyuki29@gmail.comE<gt>

=head1 SEE ALSO

Google Authenticator: http://code.google.com/p/google-authenticator/

L<Auth::GoogleAuthenticator>

L<Auth::OATH::OTP>

=cut


lib/Auth/OATH/OTP.pm
package Auth::OATH::OTP;
use strict;
use warnings;
our $VERSION = '0.0001';
use 5.008001;
use Carp ( );
use Digest::SHA ( );
use Math::Int64 ( );
use Class::Accessor::Lite (
    new => 0,
    ro  => [qw/ algorithm digits timestep /],
);

sub new
{
    my $class = shift;
    my %args  = scalar @_ == 1 && ref $_[0] eq 'HASH' ? %{ $_[0] } : @_;

    if (defined $args{digits} && $args{digits} < 6) {
        Carp::croak "'digits' must be at least 6 characters";
    }

    return bless +{
        algorithm => 'sha1',
        digits    =>  6,
        timestep  => 30,
        %args,
    }, $class;
}

sub totp
{
    my ($self, $secret, $time) = @_;
    my $bincode = pack_uint64($self->timestamp($time));
    return $self->compute_code($secret, $bincode);
}

sub hotp
{
    my ($self, $secret, $counter) = @_;
    my $bincode = pack_uint64($counter);
    return $self->compute_code($secret, $bincode);
}

sub timestamp
{
    my ($self, $time) = @_;
    $time = time unless defined $time;
    return int $time / $self->timestep;
}

sub pack_uint64
{
    return Math::Int64::uint64_to_net("$_[0]");
}

sub compute_code
{
    my ($self, $secret, $bincode) = @_;
    my @hash   = unpack 'C*', $self->hmac($secret, $bincode);
    my $offset = $hash[-1] & 0xF;
    my $digits = $self->digits;

    return sprintf "%0${digits}u", (
        ($hash[$offset + 0] & 0x7F) << 24 |
        ($hash[$offset + 1] & 0xFF) << 16 |
        ($hash[$offset + 2] & 0xFF) <<  8 |
        ($hash[$offset + 3] & 0xFF)
    ) % 10 ** $digits;
}

sub hmac
{
    my ($self, $secret, $bincode) = @_;
    my $algorithm = $self->algorithm;
    $algorithm =~ s/\D+//;

    if (my $digest = Digest::SHA->can("hmac_sha${algorithm}")) {
        return $digest->($bincode, $secret);
    }

    Carp::croak "Unknown algorithm 'HMAC-SHA${algorithm}'";
}

1;

__END__

=encoding utf-8

=head1 NAME

Auth::OATH::OTP - OATH One-Time Passcode Generator

=head1 SYNOPSIS

    use Auth::OATH::OTP;
    my $oath = Auth::OATH::OTP->new;
    my $totp = $oath->totp('MySecretPassword');
    my $hotp = $oath->hotp('MyOtherSecretPassword', 1);

=head1 DESCRIPTION

Auth::OATH::OTP is implementation of the HOTP (RFC 4226) and TOTP (RFC 6238)
algorithms as defined by OATH. (http://www.openauthentication.org)

=head1 METHODS

=head2 B<< Auth::OATH::OTP->new(%options) >>

    my $oath = Auth::OATH::OTP->new(
        algorithm => 'sha1',
        digits    =>  6,
        timestep  => 30,
    );

=over

=item B<algorithm>

Allowed values for I<algorithm> are 1, 224, 256, 384, 512, 512224, or 512256.
For more information about this parameter, see also L<Digest::SHA>.

If the argument is missing, SHA-1 will be used by default.

=item B<digits>

C<totp( )> and C<hotp( )> both default to returning 6 digits.

=item B<timestep>

I<timestep> only applies to the C<totp( )> function.

By default, the I<timestep> is 30 seconds, so there is a new password every 30 seconds.

=back

=head2 B<< $oath->totp($secret_key[, $time]) : String >>

    my $secret = pack 'A*', 'My Secret Key';
    my $totp   = $oath->totp($secret);
    # or
    my $totp   = $oath->totp($secret, time);

Returns a text string with the Time-based One Time Passcode.

Manual I<time> is an optional parameter.
If it is missing, the current time will be used by default.
This is useful for testing purposes.

=head2 B<< $oath->hotp($secret_key, $counter) : String >>

    my $secret = pack 'A*', 'My Secret Key';
    my $hotp   = $oath->hotp($secret, 1);

Returns a text string with the HMAC-based One Time Passcode.

Both parameters are required.

=head1 LICENSE

Copyright (C) hatyuki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

hatyuki E<lt>hatyuki29@gmail.comE<gt>

=head1 SEE ALSO

HOTP: L<http://www.ietf.org/rfc/rfc4226.txt>

TOTP: L<http://www.ietf.org/rfc/rfc6238.txt>

L<Authen::OATH>

=cut
